src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs

```diff
@@ -413,6 +414,270 @@ public sealed class SocketsHttpHandler_ConnectionUpgrade_Test : HttpClientTestBa
+                            const string goodbyeServer = "goodbye server";
+                            const string goodbyeClient = "goodbye client";
+
+                            clientWriter.WriteLine(helloServer);
+                            Assert.Equal(helloServer, serverReader.ReadLine());
+                            serverWriter.WriteLine(helloClient);
+                            Assert.Equal(helloClient, clientReader.ReadLine());
+                            clientWriter.WriteLine(goodbyeServer);
+                            Assert.Equal(goodbyeServer, serverReader.ReadLine());
+                            serverWriter.WriteLine(goodbyeClient);
+                            Assert.Equal(goodbyeClient, clientReader.ReadLine());
+                        }
+                    });
+                }
+            });
+        }
+
+        [Fact]
+        public async Task ConnectMethod_Fails()
+        {
+            await LoopbackServer.CreateServerAsync(async (server, url) =>
+            {
+                using (HttpClient client = CreateHttpClient())
+                {
+                    HttpRequestMessage request = new HttpRequestMessage(new HttpMethod("CONNECT"), url);
+                    request.Headers.Add(HttpKnownHeaderNames.Host,
+                                        String.Format("{0}:{1}", request.RequestUri.IdnHost, request.RequestUri.Port));
+                    // We need to use ResponseHeadersRead here, otherwise we will hang trying to buffer the response body.
+                    Task<HttpResponseMessage> responseTask = client.SendAsync(request,  HttpCompletionOption.ResponseHeadersRead);
+                    await server.AcceptConnectionAsync(async connection =>
+                    {
+                        Task<List<string>> serverTask = connection.ReadRequestHeaderAndSendCustomResponseAsync(
+                            $"HTTP/1.1 403 Forbidden\r\nDate: {DateTimeOffset.UtcNow:R}\r\nContent-Length: 7\r\n\r\nerror\r\n");
+
+                        await TestHelper.WhenAllCompletedOrAnyFailed(responseTask, serverTask);
+                        HttpResponseMessage response = await responseTask;
+
+                        Assert.True(response.StatusCode ==  HttpStatusCode.Forbidden);
+                    });
+                }
+            });
+        }
+
+        [Fact]
+        public async Task ConnectMethod_Proxy()
+        {
+            LoopbackServer.Options options = new LoopbackServer.Options { UseSsl = true };
+            HttpClientHandler handler = CreateHttpClientHandler();
+
+            var listener = new TcpListener(IPAddress.Loopback, 0);
+            listener.Start();
+            var ep = (IPEndPoint)listener.Server.LocalEndPoint;
+            Uri proxyUrl = new Uri($"http://{ep.Address}:{ep.Port}/");
+
+            //Task proxy = runProxy(listener);
+            Task proxt = LoopbackGetRequestHttpProxy.StartAsync(listener, false, false);
```

Comment:
Is it possible to use LoopbackServer here also?  I haven't looked at the LoopbackGetRequestHttpProxy stuff in detail, but generally I'd like to see us move toward using LoopbackServer unless there's a good reason not to.

Reply:
I can take a look. The LoopbackGetRequestHttpProxy already existed. I can either port functions to Loopback or mimic your last changes or simply inherit from it.  
Would it be OK to do it as separate PR? Im going to write more tests for proxy authentication. 
(and leverage work done by @Priya91 )

Reply:
Yeah, that's fine.